### PR TITLE
Bug 1560260 - fixing cards showing with nothing after system_tick

### PR DESCRIPTION
**Pull request by ScottDowne**

### CODE REVIEW

File: lib/DiscoveryStreamFeed.jsm [hunk 1]
```diff
@@ -851,9 +851,10 @@ this.DiscoveryStreamFeed = class DiscoveryStreamFeed {
    */
   async refreshAll(options = {}) {
     const { updateOpenTabs, isStartup } = options;
+    const dispatchQueue = [];
     const dispatch = updateOpenTabs
       ? action => this.store.dispatch(ac.BroadcastToContent(action))
-      : this.store.dispatch;
+      : e => dispatchQueue.push(e);
```
Comment (Mardak) [inline lib/DiscoveryStreamFeed.jsm]: Let's keep the naming consistent, so `e` -> `action` and could probably just move it to a single function, e.g.,
```js
const dispatch = action => updateOpenTabs ? store…(action) : queue…(action);
```

File: lib/DiscoveryStreamFeed.jsm [hunk 2]
```diff
@@ -851,9 +851,10 @@ this.DiscoveryStreamFeed = class DiscoveryStreamFeed {
    */
   async refreshAll(options = {}) {
     const { updateOpenTabs, isStartup } = options;
+    const dispatchQueue = [];
```
Comment (Mardak) [inline lib/DiscoveryStreamFeed.jsm]: Add a comment why we have a queue (probably something along the lines of simulating a transaction)

File: lib/DiscoveryStreamFeed.jsm [hunk 3]
```diff
@@ -865,6 +866,9 @@ this.DiscoveryStreamFeed = class DiscoveryStreamFeed {
         Cu.reportError(`Error trying to load component feeds: ${error}`)
       ),
     ]);
+    if (!updateOpenTabs) {
+      dispatchQueue.forEach(e => this.store.dispatch(e));
```
Comment (Mardak) [inline lib/DiscoveryStreamFeed.jsm]: No need for the `if` as the queue is empty otherwise. Also consistent naming of `e` -> `action`

Comment (ScottDowne) [reply lib/DiscoveryStreamFeed.jsm]: Ah I see, yes, that makes sense.

File: lib/DiscoveryStreamFeed.jsm (hunk 2 above)
Comment (ScottDowne) [reply lib/DiscoveryStreamFeed.jsm]: :+1:

File: lib/DiscoveryStreamFeed.jsm (hunk 1 above)
Comment (ScottDowne) [reply lib/DiscoveryStreamFeed.jsm]: :+1:

**Comment by Mardak** [comment]
The failing test is checking if dispatch returns the action, but it should probably check if this.store.dispatch was called with the appropriate non-broadcast action values.

File: lib/DiscoveryStreamFeed.jsm (hunk 1 above)
Comment (ScottDowne) [reply lib/DiscoveryStreamFeed.jsm]: You have a huge opinion if I keep it as two functions?

```
    const dispatch = updateOpenTabs
      ? action => this.store.dispatch(ac.BroadcastToContent(action))
      : action => dispatchQueue.push(action);
``

Comment (Mardak) [reply lib/DiscoveryStreamFeed.jsm]: Either is fine. One saves an `action =>` 9 characters 😛 whatever you think is easier to read

**Comment by Mardak** [comment]
Do we want a different approach with a narrow fix to uplift to 69 beta?

**Comment by ScottDowne** [comment]
Closing for now in favour of https://github.com/mozilla/activity-stream/pull/5246

**Pull request closed without merging**